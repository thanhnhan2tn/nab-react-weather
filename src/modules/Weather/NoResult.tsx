import React from "react";

const NoResult = () => (
  <p>No result found. Please try again</p>
);

export default NoResult;