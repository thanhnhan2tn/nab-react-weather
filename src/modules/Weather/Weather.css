.app {
  padding: 2rem 1rem;
  display: flex;
  flex-direction: column;
}
.card {
  padding: 1.2rem 3rem 1rem 3rem;
  margin: 1rem;
  box-shadow: 5px 8px 10px #80808033;
}

.img-fluid {
  float: right;
  width: 10rem;
  display: flex;
  align-items: center;
}

.focast-list .row {
  justify-content: center;
}

.row1 {
  font-size: 1.4rem;
  color: black;
}

.row3 {
  font-size: 1.4rem;
  font-weight: bold;
  color: black;
}

.focast-list .row3 {
  margin: 5% 0 2.5rem;
}

.row2 .img-fluid {
  margin: 10% 0;
  opacity: 0.5;
  width: 7rem;
}


@media (max-width: 767px) {
  .card {
    height: 10rem;
    padding: 1.5rem 1.5rem 1rem 1.5rem;
  }
  .row2 .img-fluid {
    margin: 10% 0;
    opacity: 0.5;
    width: 4rem;
  } 
  .row3 {
    font-size: 0.9rem;
  }
  .row1 {
    font-size: 0.7rem;
  }
}