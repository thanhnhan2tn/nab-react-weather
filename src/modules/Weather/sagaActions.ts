import { createAction } from "@reduxjs/toolkit";

export const loadLocation = createAction(
  "weather/getLocationAsync",
  function prepare(search: string) {
    return {
      payload: search,
    };
  }
);
export const loadWeather = createAction(
  "weather/loadWeatherAsync",
  function prepare(data: any) {
    return {
      payload: data,
    };
  }
);