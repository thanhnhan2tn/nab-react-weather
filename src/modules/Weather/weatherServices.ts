import axios from "axios";
import { API } from '../../config';

export const getWeatherService = async (id: string) => {
  const { data } = await axios.get(`${API}/weather/${id}`);
  return data;
}

export const getLocationService = async (value: string) => {
  const { data } = await axios.get(`${API}/weather?query=${value}&endpoint=search`);
  return data;
}