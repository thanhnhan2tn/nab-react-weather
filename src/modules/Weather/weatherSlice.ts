import { createSlice } from "@reduxjs/toolkit";
import { RootState } from "../../store";

interface ILocation {
  title: string;
  location_type: string;
  woeid: number;
  latt_long: string;
}

interface IWeatherState {
  loading?: boolean;
  error?: string;
  locations: ILocation[];
  weather?: unknown;
}

const initialState: IWeatherState = {
  loading: false,
  error: "",
  locations: [
    {
      title: "Ho Chi Minh City",
      location_type: "City",
      woeid: 1252431,
      latt_long: "10.759180,106.662498",
    },
  ],
  weather: {},
};

export const weatherSlice = createSlice({
  name: "location",
  initialState,
  reducers: {
    loadingLocation() {
      return {
        ...initialState,
        loading: true,
        error: "",
      };
    },
    locationReceived(state: IWeatherState, action: { payload: IWeatherState }) {
      return {
        ...state,
        loading: false,
        locations: action.payload.locations,
        weather: action.payload.weather,
        error: "",
      };
    },
    loadLocationFailue(state, action: { payload: { message: string } }) {
      return {
        ...state,
        loading: false,
        error: action.payload.message,
      };
    },
  },
});

// Action creators are generated for each case reducer function
export const { loadingLocation, locationReceived, loadLocationFailue } =
  weatherSlice.actions;

export const selectWeather = (state: RootState) => state.weather.weather;
export const selectDataState = (state: RootState) => state.weather;
export const selectLocation = (state: RootState) => state.weather.locations;

export default weatherSlice.reducer;
