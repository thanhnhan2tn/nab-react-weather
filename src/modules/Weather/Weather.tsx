import React, { Suspense, ReactNode, useCallback, useEffect } from 'react';
import { FC } from 'react';
import LocationSearch from './../../components/LocationSearch';
import moment from 'moment';
import Loader from './../../components/Loader';
import ErrorBoundary from './../../components/ErrorBoundary';
import './Weather.css';
import { Card } from 'react-bootstrap';
import WeatherItem from '../../components/WeatherCard/WeatherItem';
import WeatherCard from '../../components/WeatherCard';
import { useDispatch } from 'react-redux';
import { loadLocation } from './sagaActions';
import { selectWeather, selectDataState } from './weatherSlice';
import { DEFAULT_LOCATION } from './../../config';
import { useAppSelector } from './../../hooks';
import { IWeather, IWeatherInfo } from '../../interfaces';
import { ErrorMessage } from '../../constants';
import NoResult from './NoResult';

interface IProps {
  loading?: boolean,
  error?: string,
}

const Weather: FC = () => {
  const dispatch = useDispatch();

  const renderItems = (weathers: IWeatherInfo[]) => (
    weathers.map((forecast) => {
      return (
        <div className="col" key={forecast.id}>
          <WeatherItem data={forecast} />
        </div>
      );
    })
  )

  const { consolidated_weather = [], time, title} = useAppSelector(selectWeather) as IWeather;
  const { error, loading } = useAppSelector(selectDataState) as IProps;

  const [currentWeahter, ...weathers] = consolidated_weather;
  
  useEffect(() => {
    dispatch(loadLocation(DEFAULT_LOCATION));
  }, []);

  const onSearchLocation = useCallback((value) => {
    dispatch(loadLocation(value));
  }, [loadLocation]);
  console
  return (
    <div className="app container">
      <ErrorBoundary>
        <Suspense fallback={Loader}>
          <div className="card content">
            <LocationSearch onSearch={onSearchLocation} defaultValue={DEFAULT_LOCATION} loading={loading} />
            <div className="content__body">
              {
                error === ErrorMessage.NoResult ? <NoResult /> : ''
              }
            </div>
          </div>
          <WeatherCard weatherData={{...currentWeahter, time, title}} />
          <Card className="focast-list">
            <div className="row">
              {
                renderItems(weathers)
              }
            </div>
          </Card>
        </Suspense>
      </ErrorBoundary>
    </div>
  );
}

export default Weather;