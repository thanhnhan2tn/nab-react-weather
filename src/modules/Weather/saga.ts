import logger from "../../utils/log";
import { ErrorMessage } from "./../../constants";
import { SagaIterator } from "redux-saga";
import { call, put, takeLatest } from "redux-saga/effects";
import { loadLocation, loadWeather } from "./sagaActions";
import { getLocationService, getWeatherService } from "./weatherServices";
import { loadingLocation, loadLocationFailue, locationReceived } from './weatherSlice';

function* loadLocationSaga({
  payload,
}: {
  type: typeof loadLocation.type;
  payload: string;
}): SagaIterator {
  try {
    yield put(loadingLocation());
    const response = yield call(getLocationService, payload);
    if (response.length <= 0) {
      yield put(
        loadLocationFailue({
          message: ErrorMessage.NoResult,
        })
      );
    }
    const firstFound: { woeid: number } = response[0];
    yield put(loadWeather(firstFound.woeid));
  } catch (error: any) {
    logger.error(error);
    yield put(loadLocationFailue(error));
  }
}

function* loadWeatherSaga({
  payload,
}: {
  type: typeof loadLocation.type;
  payload: string;
}): SagaIterator {
  try {
    const weatherData = yield call(getWeatherService, payload);
    yield put(
      locationReceived({
        locations: weatherData,
        weather: weatherData.data,
      })
    );
  } catch (error: any) {
    logger.error(error);
    yield put(loadLocationFailue(error));
  }
}

export function* weatherSaga() {
  yield takeLatest(loadLocation.type, loadLocationSaga);
  yield takeLatest(loadWeather.type, loadWeatherSaga);
}