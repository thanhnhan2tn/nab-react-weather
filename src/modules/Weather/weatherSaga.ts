import { SagaIterator } from "redux-saga";
import { call, put, takeLatest } from "redux-saga/effects";

import logger from "../../utils/log";
import { ErrorMessage } from "../../constants";
import { loadLocation, loadWeather } from "./sagaActions";
import { getLocationService, getWeatherService } from "./weatherServices";
import { loadingLocation, loadLocationFailue, locationReceived, weatherReceived } from './weatherSlice';

export function* loadLocationSaga({
  payload,
}: {
  type: typeof loadLocation.type;
  payload: string;
}): SagaIterator {
  try {
    yield put(loadingLocation());
    const { data } = yield call(getLocationService, payload);
    const firstFound: { woeid: number } = data[0];
    yield put(locationReceived(data));
    yield put(loadWeather(firstFound.woeid));
  } catch (error: any) {
    logger.warn(error);
    yield put(loadLocationFailue({ message: ErrorMessage.ServerError}));
  }
}

export function* loadWeatherSaga({
  payload,
}: {
  type: typeof loadWeather.type;
  payload: string;
}): SagaIterator {
  try {
    const { data } = yield call(getWeatherService, payload);
    yield put(
      weatherReceived(data)
    );
  } catch (error: any) {
    logger.warn(error);
    yield put(loadLocationFailue({ message: ErrorMessage.ServerError}));
  }
}

export function* weatherSaga() {
  yield takeLatest(loadLocation.type, loadLocationSaga);
  yield takeLatest(loadWeather.type, loadWeatherSaga);
}