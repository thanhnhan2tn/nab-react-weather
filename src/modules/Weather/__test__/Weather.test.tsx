import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Weather from '..';
import * as hooks from '../../../hooks';

const mockStore = configureStore([]);

describe('Weather', () => {
  let store: any;
  let spyOnUseSelector: any;
  beforeEach(() => {
    store = mockStore({
      loading: false,
      error: '',
      weather: {
        consolidated_weather: [
          {
            id: 6135596511657984,
            weather_state_name: 'Heavy Cloud',
            weather_state_abbr: 'hc',
            applicable_date: '2022-01-01',
            min_temp: 21.619999999999997,
            max_temp: 31.939999999999998,
            the_temp: 30.425,
          },
          {
            id: 5195447531470848,
            weather_state_name: 'Heavy Cloud',
            weather_state_abbr: 'hc',
            applicable_date: '2022-01-02',
            min_temp: 23.285,
            max_temp: 32.485,
            the_temp: 31.155
          },
          {
            id: 4885473165049856,
            weather_state_name: 'Showers',
            weather_state_abbr: 's',
            applicable_date: '2022-01-03',
            min_temp: 23.450000000000003,
            max_temp: 30.345,
            the_temp: 28.990000000000002
          },
          {
            id: 4716299344150528,
            weather_state_name: 'Heavy Cloud',
            weather_state_abbr: 'hc',
            applicable_date: '2022-01-04',
            min_temp: 22.975,
            max_temp: 32.035,
            the_temp: 29.56
          },
          {
            id: 6218995012206592,
            weather_state_name: 'Showers',
            weather_state_abbr: 's',
            applicable_date: '2022-01-05',
            min_temp: 23.085,
            max_temp: 32.504999999999995,
            the_temp: 29.765
          },
          {
            id: 6200749722697728,
            weather_state_name: 'Heavy Cloud',
            weather_state_abbr: 'hc',
            applicable_date: '2022-01-06',
            min_temp: 23.674999999999997,
            max_temp: 32.83,
            the_temp: 29.19
          }
        ],
        time: '2022-01-01T14:35:59.233483+07:00',
        sun_rise: '2022-01-01T06:11:53.101222+07:00',
        sun_set: '2022-01-01T17:42:05.306416+07:00',
        timezone_name: 'LMT',
        title: "Ho Chi Minh City",
        location_type: "City",
        woeid: 1252431,
        latt_long: "10.759180,106.662498",
        timezone: "Asia/Ho_Chi_Minh",
      },
    });
    spyOnUseSelector = jest.spyOn(hooks, 'useAppSelector');
    spyOnUseSelector.mockReturnValue({
      consolidated_weather: [
        {
          id: 6135596511657984,
          weather_state_name: 'Heavy Cloud',
          weather_state_abbr: 'hc',
          applicable_date: '2022-01-01',
          min_temp: 21.619999999999997,
          max_temp: 31.939999999999998,
          the_temp: 30.425,
        },
        {
          id: 5195447531470848,
          weather_state_name: 'Heavy Cloud',
          weather_state_abbr: 'hc',
          applicable_date: '2022-01-02',
          min_temp: 23.285,
          max_temp: 32.485,
          the_temp: 31.155
        },
        {
          id: 4885473165049856,
          weather_state_name: 'Showers',
          weather_state_abbr: 's',
          applicable_date: '2022-01-03',
          min_temp: 23.450000000000003,
          max_temp: 30.345,
          the_temp: 28.990000000000002
        },
        {
          id: 4716299344150528,
          weather_state_name: 'Heavy Cloud',
          weather_state_abbr: 'hc',
          applicable_date: '2022-01-04',
          min_temp: 22.975,
          max_temp: 32.035,
          the_temp: 29.56
        },
        {
          id: 6218995012206592,
          weather_state_name: 'Showers',
          weather_state_abbr: 's',
          applicable_date: '2022-01-05',
          min_temp: 23.085,
          max_temp: 32.504999999999995,
          the_temp: 29.765
        },
        {
          id: 6200749722697728,
          weather_state_name: 'Heavy Cloud',
          weather_state_abbr: 'hc',
          applicable_date: '2022-01-06',
          min_temp: 23.674999999999997,
          max_temp: 32.83,
          the_temp: 29.19
        }
      ],
      time: '2022-01-01T14:35:59.233483+07:00',
      sun_rise: '2022-01-01T06:11:53.101222+07:00',
      sun_set: '2022-01-01T17:42:05.306416+07:00',
      timezone_name: 'LMT',
      title: "Ho Chi Minh City",
      location_type: "City",
      woeid: 1252431,
      latt_long: "10.759180,106.662498",
      timezone: "Asia/Ho_Chi_Minh",
    });
  });

  it('should render match  snapshot', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Weather />
      </Provider>,
    );
    const snapshot = component.toJSON();
    expect(snapshot).toMatchSnapshot();
  })
})
