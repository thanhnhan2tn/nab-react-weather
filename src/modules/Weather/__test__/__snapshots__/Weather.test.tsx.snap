// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather should render match  snapshot 1`] = `
<div
  className="app container"
>
  <div
    className="card content"
  >
    <form
      className="search"
      onSubmit={[Function]}
    >
      <input
        className="search__input"
        name="search"
        onChange={[Function]}
        placeholder="eg: Ho Chi Minh"
        type="search"
        value="Ho Chi Minh"
      />
      <button
        className="search__button"
        type="submit"
      >
        Search
        
      </button>
    </form>
    <div
      className="content__body"
    >
      
    </div>
  </div>
  <div
    className="card-today card"
  >
    <div
      className="row"
    >
      <div
        className="col-6"
      >
        <div
          className="date"
        >
          Date: 
          01/01/2022
        </div>
        <div
          className="temp"
        >
          30
           °
        </div>
        <div
          className="location"
        >
          Ho Chi Minh City
        </div>
      </div>
      <div
        className="col-6 justify-content-right"
      >
        <img
          alt="Heavy Cloud"
          className="img-fluid"
          src="https://www.metaweather.com/static/img/weather/hc.svg"
        />
      </div>
    </div>
  </div>
  <div
    className="focast-list card"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div
          className="row row1"
        >
          Min: 
          23
           °
        </div>
        <div
          className="row row1"
        >
          Max: 
          32
           °
        </div>
        <div
          className="row row2"
        >
          <img
            alt="Heavy Cloud"
            className="img-fluid"
            src="https://www.metaweather.com/static/img/weather/hc.svg"
          />
        </div>
        <div
          className="row row3"
        >
          Sunday
        </div>
      </div>
      <div
        className="col"
      >
        <div
          className="row row1"
        >
          Min: 
          23
           °
        </div>
        <div
          className="row row1"
        >
          Max: 
          30
           °
        </div>
        <div
          className="row row2"
        >
          <img
            alt="Showers"
            className="img-fluid"
            src="https://www.metaweather.com/static/img/weather/s.svg"
          />
        </div>
        <div
          className="row row3"
        >
          Monday
        </div>
      </div>
      <div
        className="col"
      >
        <div
          className="row row1"
        >
          Min: 
          23
           °
        </div>
        <div
          className="row row1"
        >
          Max: 
          32
           °
        </div>
        <div
          className="row row2"
        >
          <img
            alt="Heavy Cloud"
            className="img-fluid"
            src="https://www.metaweather.com/static/img/weather/hc.svg"
          />
        </div>
        <div
          className="row row3"
        >
          Tuesday
        </div>
      </div>
      <div
        className="col"
      >
        <div
          className="row row1"
        >
          Min: 
          23
           °
        </div>
        <div
          className="row row1"
        >
          Max: 
          33
           °
        </div>
        <div
          className="row row2"
        >
          <img
            alt="Showers"
            className="img-fluid"
            src="https://www.metaweather.com/static/img/weather/s.svg"
          />
        </div>
        <div
          className="row row3"
        >
          Wednesday
        </div>
      </div>
      <div
        className="col"
      >
        <div
          className="row row1"
        >
          Min: 
          24
           °
        </div>
        <div
          className="row row1"
        >
          Max: 
          33
           °
        </div>
        <div
          className="row row2"
        >
          <img
            alt="Heavy Cloud"
            className="img-fluid"
            src="https://www.metaweather.com/static/img/weather/hc.svg"
          />
        </div>
        <div
          className="row row3"
        >
          Thursday
        </div>
      </div>
    </div>
  </div>
</div>
`;
